core/trino-main/src/main/java/io/trino/operator/ChangeOnlyUpdatedColumnsMergeProcessor.java

```diff
@@ -83,6 +91,20 @@ public Page transformPage(Page inputPage)
 
         Page result = new Page(builder.toArray(Block[]::new));
 
+        for (int nonNullColumnChannel : nonNullColumnChannels) {
+            Block block = result.getBlock(nonNullColumnChannel);
+            if (block.mayHaveNull()) {
+                for (int position = 0; position < positionCount; position++) {
+                    long operation = TINYINT.getLong(operationChannelBlock, position);
+                    if (operation == INSERT_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                        if (block.isNull(position)) {
+                            throw new TrinoException(CONSTRAINT_VIOLATION, "Assigning NULL to non-null MERGE target table column");
```

Comment:
It would be useful to indicate which column has the issue

Reply:
Agreed, but getting the column names all the way here seems complicated and can be done as a follow up.

Reply:
We could add them to `MergeParadigmAndTypes`

Reply:
Added, and now the non-null column test lists the non-null column name assigned to.